### PR TITLE
Use Readers in fileio module

### LINKED ISSUE #2494: Align `pv.read` and `pv.get_reader`
**Describe what maintenance you would like added.**

It would be ideal to have `pv.read` be a shortcut for `pv.get_reader`.  This is a super simplified example, but in the most straightforward case, it would look like this:

```py
def read(filename):
    reader = pv.get_reader(filename)
    return reader.read()
```

Then we only need to maintain vtk readers in one location, while maintaining the one-liner of `pv.read()` as compared to the two liner above.

### Complexities

- Instead of allowing users to use the `attrs` argument to `pv.read`, it would be better to use the Reader class where we expose methods and attributes.  We should deprecate or remove this argument.
- Not all readers in `pv.read` are implemented as a class.  Here is a checklist of missing Reader classes:
- [ ] vtkBMPReader
- [ ] vtkDEMReader
- [ ] vtkJPEGReader
- [ ] vtkMetaImageReader
- [ ] vtkNrrdReader
- [ ] vtkPNGReader
- [ ] vtkPNMReader  <- not tested
- [ ] vtkSLCReader
- [ ] vtkTIFFReader
- [ ] vtkGLTFReader
- [ ] vtkHDRReader
- [ ] vtkFluentReader  <- not tested
- [ ] vtkMFIXReader  <- not tested
- [ ] vtkPTSReader
- [ ] vtkAVSucdReader
- [ ] vtkSegYReader
- [ ] vtkHDFReader

Seperate implementations in fileio that do not exist in Readers classes:
- [ ] vtkExodusIIReader

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

Closes #2494.  This PR uses the machinery in the `pyvista.utilities.reader` module to replace the functionality that the `standard_reader_routine` did.

This enables one location to maintain reading utilities.

### Details

I will leave `read_exodus` out of this PR as there is no Reader class and it is self-contained in fileio.

TODO:
- [x] implement `read_plot3d` functionality into the Reader
- [x] ~~Add error observing functionality into BaseReader?~~ Continue wrapping observer for `pyvista.read`
- [x] Eliminate `standard_reader_routine`

### CODE REVIEW

**Comment by whophil** [comment]
Hey @MatthewFlamm , sorry for the delay. I haven't been keeping up with the changes you've been making to the readers, so excuse my ignorance here.

- After your refactoring, will `read_plot3d()` still exist?
- If so, are you proposing to remove its `q_filenames` argument?

I have never used the `vtkPlot3DMetaReader`. My understanding is that the JSON format it reads is either an invention of ParaView or VTK, and I have not worked with any code that outputs it. It seems just to store the flags that are used by the actual reader (`vtkMultiBlockPLOT3DReader`). My most common use case is to just use `read_plot3d()` to read in a single grid file and a single q file.

**Comment by MatthewFlamm** [comment]
> * After your refactoring, will `read_plot3d()` still exist?

Im not sure, that's what I wanted to discuss.  I'm proposing that all the other routines other than `pyvista.read` will be deprecated and/or removed.  It will be suggested to use the class based readers for cases where more control is needed, or where inspection of the reader state is needed, like:

```py
reader = Plot3DMetaReader(filename)
reader.add_q_files(q_file)
mesh = reader.read()
```

After looking into this, it isn't clear to me that adding q-files is intended to be user-facing.  It has a different method name in 8.x, 9.0.x, and 9.1.x.  I'm unsure how we can get the functionality to more closely resemble the json file format without a lot of effort, which from my reading of the vtk docs is the intended usage due this reader. Maybe the best solution is to replicate the q file usage in the class based Reader, then someone with more need can make it more elegant?

**Comment by whophil** [comment]
I believe the thing that should be wrapped, and possibly exposed with a convenience class (like the CGNS reader) is actually `vtkMultiBlockPLOT3DReader` and not `vtkPlot3DMetaReader`. Regardless, the ability to read Plot3d data with Q files is in my opinion a necessity. I'm happy to help with the Plot3D refactor, as I  likely use Plot3D file formats more than you.

**Comment by MatthewFlamm** [comment]
> believe the thing that should be wrapped, and possibly exposed with a convenience class (like the CGNS reader) is actually `vtkMultiBlockPLOT3DReader` and not `vtkPlot3DMetaReader`

You have uncovered something I overlooked.  `pyvista.read` uses  `vtkPlot3DMetaReader` while `read_plot_3d` uses `vtkMultiBlockPLOT3DReader`.  This clears up a lot of my confusion.  I think we can add a new reader for `vtkMultiBlockPLOT3DReader` with the q file functionality.

File: pyvista/utilities/fileio.py
Comment (MatthewFlamm) [inline pyvista/utilities/fileio.py:140]: This seems weird and the commit in #495 that adds this references Python 3.5.  Can we remove this except?

Comment (adeak) [reply pyvista/utilities/fileio.py:140]: Good catch, yes please! Seems like `meshio` was using f-strings while <3.6 was supported in PyVista. We don't even have 3.6 now, this should go.

Comment (MatthewFlamm) [reply pyvista/utilities/fileio.py:140]: Good, this allows for more cleanup here (and allows a higher % of coverage of the changes).

**Comment by MatthewFlamm** [comment]
This is now fully ready for review.

I fully removed `read_legacy`, but we could consider deprecating it instead.

There aren't tests for all the file types, so it would be nice if someone can review the mapping of readers based on extension in the old `fileio` module with the mapping in `reader` module to make sure I didn't make any typos/mistakes on the extensions (or missing like `.img` was missing).

File: pyvista/utilities/fileio.py
Comment (adeak) [inline pyvista/utilities/fileio.py:332]: If we're deprecating this then I wonder if we should do the same to `read_legacy()` after all.

And the updated code below (and emitted warning) uses MultiBlockPlot3Reader; shouldn't this deprecation note also point to that?

**Review by adeak — COMMENTED** [review]
I can only look at the meat of the changes later, so only a few cursory remarks for now.

File: pyvista/utilities/fileio.py
Comment (MatthewFlamm) [reply pyvista/utilities/fileio.py:332]: > And the updated code below (and emitted warning) uses MultiBlockPlot3Reader; shouldn't this deprecation note also point to that?

Good catch!

Comment (MatthewFlamm) [reply pyvista/utilities/fileio.py:332]: > If we're deprecating this then I wonder if we should do the same to `read_legacy()` after all.

You've convinced me to deprecate it instead, it isn't too much extra work.

File: doc/api/utilities/utilities.rst
Comment (MatthewFlamm) [reply doc/api/utilities/utilities.rst:36]: We should either deprecate or remove `read_legacy` IMO.  Even before this PR when you called `pyvista.read('file.vtk')` it called `read_legacy`, set there was no need to have a separate function.

Exodus currently does not have a Reader class and it is self contained in that it doesn't use `standard_reader_routine`.  So I specifically left it out if this PR to keep things manageable.

File: tests/utilities/test_reader.py
Comment (MatthewFlamm) [reply tests/utilities/test_reader.py:352]: A better question for @whophil. IMO this test is more comprehensive than the prior test for this reader since we are calling methods on the vtk object.

Comment (whophil) [reply tests/utilities/test_reader.py:352]: Perhaps add a call to `reader.read()` after calling `add_q_files()` and assert that point arrays are loaded? (It does work). The test is certainly better than it was before.

Besides that, I don't think we're normally in the business of testing every permutation of the VTK reader class.

**Comment by whophil** [comment]
Two thoughts:
1. Eventually `MultiBlockPlot3DReader` should have properties for all of the useful reader flags. The ParaView property window for `vtkMultiBlockPlot3DReader` is shown below. Not having these does not block this PR, in my opinion.
    image
2. `test_reader::test_multiblockplot3dreader` reports the following to stdout. The data reads fine, but the presence of the error might mislead users.
    ```
    2022-05-12 00:06:15.803 (   0.525s) [          BF9E13]vtkMultiBlockPLOT3DRead:2555   ERR| vtkMultiBlockPLOT3DReader (0x7fab5381c6f0): The settings that you provided do not match what was auto-detected in the file. Set `ForceRead` to true to get past this error.
    The detected settings are: 
    BinaryFile: 1
    ByteOrder: 1
    HasByteCount: 1
    MultiGrid: 1
    NumberOfDimensions: 3
    DoublePrecision: 1
    IBlanking: 0
    ```
    This error is issued because `vtkMultiBlockPlot3DReader` appears to "test" the settings of the reader against the auto-detected settings when `UpdateInformation()` is called, but only if auto-detect is `False`.
    
    One solution would be to set auto-detect to `True` by default, and to rework `BaseReader` so that attribute can be correctly set before the call to `UpdateInformation()`, which is triggered by the `BaseReader.__directory` setter. Alternately, `MultiBlockPlot3DReader.__init__` could just override `BaseReader.__init__`, but that seems less maintainable.

**Comment by MatthewFlamm** [comment]
> 1. Eventually `MultiBlockPlot3DReader` should have properties for all of the useful reader flags.

I agree.  This PR should make it much easier to add those in the future.  This PR adds the existing functionality.

> test_reader::test_multiblockplot3dreader reports the following to stdout.

Good catch!  I refactored the `__init__` of `BaseReader` so that we no longer need to override the whole `__init__` method to set defaults.  I used this in this reader and the other Readers.

File: tests/utilities/test_reader.py
Comment (MatthewFlamm) [reply tests/utilities/test_reader.py]: Now we test the arrays are not loaded without q file and are loaded with q file.

**Review by akaszynski — APPROVED** [review]
This is a way cleaner implementation, and I fully support moving to use our new `get_reader` approach.

I see no issues with this and I'd like to get this merged to `main` and start playing around with this locally on some of my closed source projects to make sure that the new reader works. I'm happy with this as-is.

File: pyvista/utilities/fileio.py
Comment (MatthewFlamm) [inline pyvista/utilities/fileio.py:187]:
```suggestion
```

**Comment by MatthewFlamm** [comment]
This is a fairly large PR.  I'll give it just a few more days in case there are any more comments.  I agree that testing downstream will also be important as there wasn't comprehensive testing of file types in `fileio`.

**Comment by akaszynski** [comment]
> I'll give it just a few more days in case there are any more comments.

Agreed. I'll merge by Friday next week unless there are additional concerns raised.

**Pull request merged**
